src/Widgets/TerminalWidget.vala

```diff
@@ -261,22 +335,183 @@ namespace Terminal {
+                    }
+                }
+
+                return false;
+            }
+
+            if (CONTROL_MASK in modifiers && Application.settings.get_boolean ("natural-copy-paste")) {
+                if (match_keycode (Gdk.Key.c, keycode)) {
+                    if (get_has_selection ()) {
+                        copy_clipboard ();
+                        if (!(SHIFT_MASK in modifiers)) { // Shift not pressed
+                            unselect_all ();
+                        }
+                        return true;
+                    } else {
+                        last_key_was_return = true; // Ctrl-c: Command cancelled
+                    }
+                } else if (match_keycode (Gdk.Key.v, keycode) && clipboard.wait_is_text_available ()) {
+                    paste_clipboard ();
+                    return true;
+                }
+            }
+
+            if (MOD1_MASK in modifiers) {
+                if (keyval == Gdk.Key.Up) {
+                    return !scroll_to_command_action.enabled;
+                } else if (match_keycode (Gdk.Key.c, keycode)) {
+                    var has_output = !resized && get_last_output ().length > 0;
+                    copy_output_action.set_enabled (has_output);
+                }
             }
 
             return false;
         }
 
+        private void setup_menu () {
+            // Update the "Paste" menu option
+            clipboard.request_targets ((clipboard, atoms) => {
+                bool can_paste = false;
+
+                if (atoms != null && atoms.length > 0) {
+                    can_paste = Gtk.targets_include_text (atoms) || Gtk.targets_include_uri (atoms);
+                }
+
+                paste_action.set_enabled (can_paste);
+            });
+
+            // Update the "Copy Last Outptut" menu option
+            var has_output = !resized && get_last_output ().length > 0;
+            copy_output_action.set_enabled (has_output);
+        }
+
+        protected override void copy_clipboard () {
+            if (link_uri != null && !get_has_selection ()) {
+                clipboard.set_text (link_uri, link_uri.length);
+            } else {
+                base.copy_clipboard ();
+            }
+        }
+
+        private void copy_output () {
+            var output = get_last_output ();
+            clipboard.set_text (output, output.length);
+        }
+
+        protected override void paste_clipboard () {
+            clipboard.request_text ((clipboard, text) => {
+                if (text == null) {
+                    return;
+                }
+
+                if (!text.validate ()) {
+                    warning ("Dropping invalid UTF-8 paste");
+                    return;
+                }
+
+                unowned var toplevel = (MainWindow) get_toplevel ();
+
+                if (!toplevel.unsafe_ignored && Application.settings.get_boolean ("unsafe-paste-alert")) {
+                    string? warning = null;
```

Comment:
Best not use the name of a command as a variable name.  Cannot insert `warning ()` commands into the code when this variable is in scope.